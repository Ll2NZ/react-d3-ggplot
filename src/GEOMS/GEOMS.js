import React            from "react";
import PropTypes        from "prop-types";
import { ChartContext } from "../_context/";

function GEOMS(props) {
        return(
            <ChartContext.Provider value={ props }>
                <svg width={ props.dimensions.width } height={ props.dimensions.height }>
                    { props.children }
                </svg>
            </ChartContext.Provider>
        );
}

GEOMS.propTypes = {
    dimensions: PropTypes.object,
    children: PropTypes.any
};

export default GEOMS;
