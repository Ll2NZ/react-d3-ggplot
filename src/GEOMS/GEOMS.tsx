import React, { ReactChildren, ReactComponentElement, ReactChild } from "react";
import { ChartContext, IContext } from "../_context/";

interface IProps extends IContext {
  children?: ReactChild[] | ReactChild;
}

function GEOMS(props: IProps) {
  if (!props.data || !props.aes || !props.dimensions) {
    throw new Error(`GEOMS must have the following properties: data, aes, and dimension`);
  }

  if (!props.children) {
    throw new Error("GEOMS expects at least one child components such as: <Rects/> or <Points/>");
  }

  return (
    <ChartContext.Provider value={props}>
      <svg width={props.dimensions.width} height={props.dimensions.height}>
        {props.children}
      </svg>
    </ChartContext.Provider>
  );
}

export default GEOMS;
