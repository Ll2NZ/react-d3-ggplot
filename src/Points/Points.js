import React           from "react";
import { Component }   from "react";
import PropTypes       from "prop-types";
import { scaleFinder } from "../Utilities/";
import { select }      from "d3-selection";

class Points extends Component {
    static defaultProps = {
         dimensions:
        {
            width: window.innerWidth*0.9,
            height: window.innerHeight*0.9,
            padding: 50
        },
        className: null,
        color: "orange",
        radius: 1
    }

    static propTypes = {
        data: PropTypes.array.isRequired,
        dimensions: PropTypes.shape({
            width: PropTypes.number,
            height: PropTypes.number,
            padding: PropTypes.number
        }),
        scaleTypes: PropTypes.array,
        className: PropTypes.string,
        aes: PropTypes.array,
        color: PropTypes.string,
        radius: PropTypes.number
    }

    getXScale(){
        // get data and x and y value pair
        const { data, aes } = this.props;

        // convert data to a scale to be used for our x-values
        const xValues  = data.map(item => item[aes[0]]);
        const scaleObj = new scaleFinder(xValues);

        return scaleObj;
    }

    getXScaleType(){
        const { dimensions, scaleTypes } = this.props;
        const scaleObj                   = this.getXScale();
        let xScale;

        // find appropiate scale
        if(scaleTypes[0] === "linear")
            xScale = scaleObj.getLinearScale().nice();

        if(scaleTypes[0] === "time")
            xScale = scaleObj.getTimeScale().nice();

        if(scaleTypes[0] === "ordinal")
            xScale = scaleObj.getOrdinalScale(0.5); // .nice() method not available

        // set scale range
        xScale.range([dimensions.padding, dimensions.width - dimensions.padding]);

        return xScale;
    }

    getYScale(){
       // get y-values
        const { data, aes } = this.props;
        const yValues       = data.map(item => item[aes[1]]);

        // create xScale
        const scaleObj  = new scaleFinder(yValues);

        return scaleObj;
    }

    getYScaleType(){
        const { scaleTypes, dimensions } = this.props;
        const scaleObj                   = this.getYScale();
        let yScale;

        if(scaleTypes[1] === "linear")
            yScale = scaleObj.getLinearScale().nice();

        if(scaleTypes[1] === "time")
            yScale = scaleObj.getTimeScale().nice();

        if(scaleTypes[1] === "ordinal")
            yScale = scaleObj.getOrdinalScale(0.5); // .nice() method not available

        // set scale range
        yScale.range([dimensions.height - dimensions.padding, dimensions.padding]);

        return yScale;
    }

    appendCircles(){
        // scales
        const xScale            = this.getXScaleType();
        const yScale            = this.getYScaleType();

        // all props needed for points
        const { color, radius } = this.props;
        const { data, aes }     = this.props;

        // clear graph for next set of data points
        if(this.node.children.length > 0)
            select(this.node).selectAll("circle").remove();

        // append new points
        select(this.node)
            .selectAll("circle")
            .data(data)
            .enter()
            .append("circle")
            .attr("cx", d => xScale(d[aes[0]]))
            .attr("cy", d => yScale(d[aes[1]]))
            .attr("r", radius)
            .attr("fill", color);
    }

    render(){
        return(
            <g
                ref={ node => this.node = node }
                className={ this.props.className }
            >
            </g>
        );
    }

    componentDidMount(){
        this.appendCircles();
    }

    componentDidUpdate(){
        this.appendCircles();
    }
}

export default Points;
