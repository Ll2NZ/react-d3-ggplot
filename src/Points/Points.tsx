import React from "react";
import { useContext } from "react";
import { ChartContext } from "../_context/";
import { useXScale } from "../_hooks/";
import { useYScale } from "../_hooks/";

interface IProps {
  r?: number;
  fill?: string;
}

Points.displayName = "Points";

function Points(props: IProps) {
  const context = useContext(ChartContext);
  const xScale: any = useXScale();
  const yScale: any = useYScale();

  // will remove later when testing componet
  if (!xScale || !yScale) {
    return <h1>whoops</h1>;
  }

  return (
    <g>
      {/*
        For the key, we're not supposed to used index because we can't take
        advantage of react's optimization capabilities when it comes to re-rendering.
        Will use key until we can sort this detail out.
      */}
      {context.data.map((datum, index) => (
        <circle
          key={index}
          cx={xScale(datum[context.aes[0]])}
          cy={yScale(datum[context.aes[1]])}
          r={!props.r ? 3 : props.r}
          fill={!props.fill ? "orange" : props.fill}
        />
      ))}
    </g>
  );
}

export default Points;
