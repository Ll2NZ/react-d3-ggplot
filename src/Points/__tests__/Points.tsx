// ALE complains when we just use import React from "react";
import * as React from "react";
import Points from "../Points";
import { GEOMS } from "../../GEOMS/";
import { render, cleanup } from "react-testing-library";

afterEach(cleanup);

const validState = {
  data: [{ x: 0, y: 0 }, { x: 10, y: 10 }],
  aes: ["x", "y"],
  dimensions: { width: 500, height: 300, padding: 50 }
};

test("Points renders a <g> element wrapping N <point> elements", () => {
  // We don't care about testing <GEOMS/> here.
  const { container } = render(
    <GEOMS {...validState}>
      <Points />
    </GEOMS>
  );

  expect(container.firstChild).toMatchSnapshot();
});
