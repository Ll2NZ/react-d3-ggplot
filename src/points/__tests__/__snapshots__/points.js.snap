// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Underlying SVG markup should not change when we change our component logic 1`] = `
<svg
  height="300"
  width="500"
>
  <g>
    <circle
      cx="50"
      cy="250"
      fill="orange"
      r="2"
    />
    <circle
      cx="450"
      cy="50"
      fill="orange"
      r="2"
    />
  </g>
</svg>
`;
